esp-hal/src/dma/mod.rs

```diff
@@ -1534,25 +1522,40 @@ pub trait DmaChannel: crate::private::Sealed {
 
     /// A description of the TX half of a DMA Channel.
     type Tx: TxRegisterAccess + InterruptAccess<DmaTxInterrupt>;
+}
 
+/// A description of a DMA Channel that can be used with a peripheral.
+#[cfg(pdma)]
+pub trait PeripheralDmaChannel: DmaChannel {
     /// A suitable peripheral for this DMA channel.
     type P: PeripheralMarker;
 }
 
 #[doc(hidden)]
 pub trait DmaChannelExt: DmaChannel {
-    type Degraded: DmaChannel;
-
     fn get_rx_interrupts() -> impl InterruptAccess<DmaRxInterrupt>;
     fn get_tx_interrupts() -> impl InterruptAccess<DmaTxInterrupt>;
 
-    fn degrade_rx(rx: Self::Rx) -> <Self::Degraded as DmaChannel>::Rx;
-    fn degrade_tx(tx: Self::Tx) -> <Self::Degraded as DmaChannel>::Tx;
-
     #[doc(hidden)]
     fn set_isr(handler: InterruptHandler);
 }
 
+#[doc(hidden)]
+pub trait DmaChannelConvert<DEG: DmaChannel>: DmaChannel {
```

Comment:
I'm wondering if `DEG` should be an associated type. Do we expect a DMA channel to degrade into multiple channel types? Or perhaps this was needed to make the compiler happy.

Reply:
This will allow me to implement `DmaChannelConvert<AnySpiPdmaChannel> for Spi3DmaChannel` and likes, which allows keeping the current "ZST if peripheral is concrete" but converts to degraded for AnySpi in the future. This was a conscious move from associated types that you can still find a few commits back.